core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java

```diff
@@ -3269,6 +3275,7 @@ && isHourKeyword(columnExpr.token)
                             configuration,
                             factory,
                             meta,
+                            executionContext.getSharedWorkerCount(),
```

Comment:
In such case, work stealing will use the extra slot with `workerCount` index to access VAFs, so I don't think I follow. Could you clarify how exactly the issue would look like?

Reply:
two HTTP connections execute SQL with async offload. HTTP has own worker pool. There are variations: http pool > shared pool and http pool < shared pool. Lets not bother with pool sizes for now.

SQL1 runs on http worker 0
SQL2 runs on http worker 1

Async offload for both runs on shared pool.

Lets assume SQL1 and SQL2 work steal simultaneously. Which `vaf` slot will they be using ?

Reply:
> Lets assume SQL1 and SQL2 work steal simultaneously. Which `vaf` slot will they be using ?

They will be using `<shared_worker_pool_size>` slot of their own VAF list. But since each record cursor uses its own VAF list, there will be no clashing/contention. Am I missing something?

Reply:
does stealing check what is it stealing ? Can SQL1 and SQL2 stealing code process SQL2 and SQL2 parts. Note that SQL1 stealing is processing SQL2 work.

Reply:
HTTP workers won't be processing each other's tasks - the tasks are handled by `GroupByJob` instances running on the shared pool. In this sense, SQL1 runs only on HTTP worker 0 and shared worker threads, and SQL2 runs only on HTTP worker 1 and shared worker threads. They only do "work stealing" when they can't publish a task, i.e. they start processing the next task locally. That's different from work stealing used in async offload, but this PR only touches what's in `.engine.groupby.vect` package.

Reply:
My bad, you're referring to `GroupByNotKeyedVectorRecordCursorFactory#getRunWhatsLeft`. I see the problem now. Let me think of a solution.

Reply:
The race should be fixed in 590971f08bb64b3428c7b2862cd6d2ebc54c863d. Thanks for spotting it!

Reply:
It turns out that the initial fix attempt was ok and no additional synchronization is necessary. That's because, unlike in async offload, VAF work stealing is done only on the thread's own tasks, so no one else will be using `workerCount` index (i.e. the extra work stealing slot). The only thing we need is the non-atomic field access fix around `AbstractLockable#tryLock()`, but that will be added by @ideoma in #3053.

Reply:
As the fix in #3072 introduced proper work stealing for vectorized GROUP BY tasks, I've reverted this patch back to async offload-like synchronization. It should be ready for review now.